evm/evm.go

```diff
@@ -0,0 +1,394 @@
+// Copyright (C) 2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+package evm
+
+import (
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ava-labs/avalanche-tooling-sdk-go/utils"
+	"github.com/ava-labs/subnet-evm/accounts/abi/bind"
+	"github.com/ava-labs/subnet-evm/core/types"
+	"github.com/ava-labs/subnet-evm/ethclient"
+	"github.com/ava-labs/subnet-evm/rpc"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+const (
+	BaseFeeFactor               = 2
+	MaxPriorityFeePerGas        = 2500000000 // 2.5 gwei
+	NativeTransferGas    uint64 = 21_000
+	repeatsOnFailure            = 3
+	sleepBetweenRepeats         = 1 * time.Second
+)
+
+func ContractAlreadyDeployed(
+	client ethclient.Client,
+	contractAddress string,
+) (bool, error) {
+	if bs, err := GetContractBytecode(client, contractAddress); err != nil {
+		return false, err
+	} else {
+		return len(bs) != 0, nil
+	}
+}
+
+func GetContractBytecode(
+	client ethclient.Client,
+	contractAddressStr string,
+) ([]byte, error) {
+	contractAddress := common.HexToAddress(contractAddressStr)
+	var (
+		code []byte
+		err  error
+	)
+	for i := 0; i < repeatsOnFailure; i++ {
+		ctx, cancel := utils.GetAPILargeContext()
+		defer cancel()
+		code, err = client.CodeAt(ctx, contractAddress, nil)
+		if err == nil {
+			break
+		}
+		err = fmt.Errorf(
+			"failure obtaining code for %s on %#v: %w",
+			contractAddressStr,
+			client,
+			err,
+		)
+		time.Sleep(sleepBetweenRepeats)
+	}
+	return code, err
+}
+
+func GetAddressBalance(
+	client ethclient.Client,
+	addressStr string,
+) (*big.Int, error) {
+	address := common.HexToAddress(addressStr)
+	var (
+		balance *big.Int
+		err     error
+	)
+	for i := 0; i < repeatsOnFailure; i++ {
```

Comment:
this code can be moved to the separate func with func param client.CodeAt or client.BalanceAt as these 2 loops are very similar to each other